### DISCUSSION TITLE
custom jwt authentication

### DISCUSSION THREAD

**Discussion by webafra** [loopbackio/loopback-next#7193] [Q&A]
Hello.

I want to use this service to develop my own personalized authentication.

1. Find the user
2. Create the JWT access token for that user

```
let findUser = await this.userRepository.findOne({
      where: {
        mobile: credentials.mobile
      },
      order: ["id DESC"]
    });

    const token = await this.jwtService.generateToken(findUser.id);

    return { access_token: token }
```

The code I wrote does not work because method (generateToken) inputs argument must be an email and a name.
I want to authenticate with a mobile number.

please help me.

**Comment by dhmlau** [comment]
@jannyHou @mrmodise, could you please help with this question? Thanks.

**Comment by jannyHou** [comment]
@webafra You can create your own JWT service like [this example](https://github.com/strongloop/loopback4-example-shopping/blob/a0f474f0946084f710de91f75ffbe0d4df690235/packages/shopping/src/services/jwt.service.ts), then bind it to binding key that `authentication-jwt` module exported as [this.bind(TokenServiceBindings.TOKEN_SERVICE).toClass(JWTService)](https://github.com/strongloop/loopback4-example-shopping/blob/a8262a9ce15fa19321f526ad9bb34a51a88d03a1/packages/shopping/src/application.ts#L101) in the app constructor .

**Comment by Arathy-sivan** [comment]
Can we create API without using model and repository in loopback4??

**Reply by infortraining** [reply]
its impossible

**Comment by mrmodise** [comment]
@Arathy-sivan very possible. The default [ping controller](https://github.com/strongloop/loopback4-example-shopping/blob/a0f474f0946084f710de91f75ffbe0d4df690235/packages/shopping/src/controllers/ping.controller.ts#L38) is a good example of such a case

**Comment by mrmodise** [comment]
> authenticate with a mobile number.

Please see @jannyHou above and let me know if you still need help. I will be glad to help

**Comment by Lipinalipi** [comment]
how can we generate token without using model and repository in loopback4?

**Comment by jannyHou** [comment]
@Lipinalipi If you don't need persistence, you can just generate token using services like https://github.com/strongloop/loopback-next/blob/master/extensions/authentication-jwt/src/services/jwt.service.ts, see function [generateToken](https://github.com/strongloop/loopback-next/blob/master/extensions/authentication-jwt/src/services/jwt.service.ts#L54).

**Reply by emazzu** [reply]
one question, lb4, has the same model that lb3, model user, role, roleMapping, acl and accessToken ???
Because I have to migrate from lb3 with these security models ??
Regards
Eduardo
